zigpy/zcl/clusters/general.py

```diff
@@ -1278,6 +1286,444 @@ class PollControl(Cluster):
+            + payload.to_bytes(payload_length, "little"),
+            payload_length + 1,
+        )
+        if calcul_mic is not None and calcul_mic != mic:
+            LOGGER.debug(
+                "Wrong mic : %s, calcul mic %s, ignore frame", hex(mic), hex(calcul_mic)
+            )
+            return
+        (
+            command,
+            type,
+            schema,
+            cluster_id,
+            zcl_command_id,
+            value,
+        ) = GreenPowerProxy.command[command_id]
+        payload, _ = t.deserialize(payload.to_bytes(payload_length, "little"), schema)
+        LOGGER.debug(
+            "Green power frame ieee : %s, command_id : %s, payload : %s,counter : %s",
+            ieee,
+            command_id,
+            payload,
+            counter,
+        )
+        value = value + tuple(payload)
+        dev = application.devices[ieee]
+        if counter is not None:
+            if (
+                0x9999
+                in dev.endpoints[zigpy.zcl.clusters.general.GreenPowerProxy.endpoint_id]
+                .in_clusters[zigpy.zcl.clusters.general.GreenPowerProxy.cluster_id]
+                ._attr_cache
+                and dev.endpoints[
+                    zigpy.zcl.clusters.general.GreenPowerProxy.endpoint_id
+                ]
+                .in_clusters[zigpy.zcl.clusters.general.GreenPowerProxy.cluster_id]
+                ._attr_cache[0x9999]
+                > counter
+            ):
+                LOGGER.debug("Already get this frame counter,I ignoring it")
+                return
+            dev.endpoints[
+                zigpy.zcl.clusters.general.GreenPowerProxy.endpoint_id
+            ].in_clusters[
+                zigpy.zcl.clusters.general.GreenPowerProxy.cluster_id
+            ]._update_attribute(
+                0x9999, counter
+            )
+        if cluster_id is not None and type == "CLUSTER_COMMAND":
+            if cluster_id not in dev.endpoints[1].out_clusters:
+                dev.endpoints[1].add_output_cluster(cluster_id)
+                application.device_initialized(dev)
+            dev.endpoints[1].out_clusters[cluster_id].handle_message(
+                foundation.ZCLHeader.cluster(
+                    application.get_sequence(), zcl_command_id
+                ),
+                value,
+            )
+
+    def handle_message(self, data):
+        data = data[0]
+        if data[0] == 12 and data[5] == 224:
+            ieee = t.EUI64(data[1:5] + data[1:5])
+            type = data[6]
+            self.create_device(ieee, type)
+            return
+        payload = ()
+        ieee = t.EUI64(data[2:6] + data[2:6])
+        counter = int.from_bytes(data[6:10], byteorder="little")
+        command_id = data[10]
+        payload_length = len(data[11:-4])
+        payload = int.from_bytes(data[11:-4], byteorder="little")
+        mic = int.from_bytes(data[-4:], byteorder="little")
+        header = int.from_bytes(data[0:2], byteorder="little")
```

Comment:
for decoding binary messages using the `struct` module might make this a bit shorter

Reply:
It's not also simple. The message structure depend of the first byte of message. For now I handle only most simple structure (not encrypt level 2 and 3).